e2e/tests/truffle/test/GoContracts.js

```diff
@@ -0,0 +1,88 @@
+const fs = require('fs');
+const path = require('path');
+const Web3 = require('web3');
+const InnerEmitter = artifacts.require('InnerEmitter');
+const OuterEmitter = artifacts.require('OuterEmitter');
+const {
+    SpeculativeNonceTxMiddleware, SignedTxMiddleware, Client,
+    LocalAddress, CryptoUtils, LoomProvider, Contracts, Address
+} = require('loom-js');
+
+
+
+
+contract('SampleGoContract', async () => {
+    const privateKey = CryptoUtils.generatePrivateKey();
+    const publicKey = CryptoUtils.publicKeyFromPrivateKey(privateKey);
+    let client, web3js;
+
+    beforeEach(async () => {
+        const nodeAddr = fs.readFileSync(path.join(process.env.CLUSTER_DIR, '0', 'node_rpc_addr'), 'utf-8').trim();
+        const chainID = 'default';
+        const writeUrl = `ws://${nodeAddr}/websocket`;
+        const readUrl = `ws://${nodeAddr}/queryws`;
+
+        client = new Client(chainID, writeUrl, readUrl);
+        client.on('error', msg => {
+            console.error('Error on connect to client', msg);
+            console.warn('Please verify if loom cluster is running')
+        });
+
+        const setupMiddlewareFn = function (client, privateKey) {
+            const publicKey = CryptoUtils.publicKeyFromPrivateKey(privateKey);
+            return [new SpeculativeNonceTxMiddleware(publicKey, client), new SignedTxMiddleware(privateKey)]
+        };
+        client.txMiddleware = setupMiddlewareFn(client, privateKey);
+
+        web3js = new Web3(new Web3.providers.HttpProvider(`http://${nodeAddr}/eth`));
+    });
+
+    it('nested event emitted from go contract', async () => {
+        let innerEmitter = await InnerEmitter.deployed();
+        let outerEmitter = await OuterEmitter.deployed();
+
+        const innerEmitterValue = 31;
+        const outerEmitterValue = 63;
+
+        const sampleGoContract = await new Contracts.SampleGoContract.createAsync(
+            client,
+            new Address(client.chainId, LocalAddress.fromPublicKey(publicKey))
+        );
+
+        const goResult = await sampleGoContract.testNestedEvmCallsAsync(
+            new Address(client.chainId, LocalAddress.fromHexString(innerEmitter.address)),
+            new Address(client.chainId, LocalAddress.fromHexString(outerEmitter.address)),
+            innerEmitterValue,
+            outerEmitterValue,
+        );
+        const goContractLogs = await web3js.eth.getPastLogs({
+            address: innerEmitter.address,
+        });
```

Comment:
So `innerEmitter.address` gets us logs for `outerEmitter` too? That seems a bit odd.
What do I get if I call `getPastLogs({ address: outerEmitter.address })`?
And what happens with `getPastLogs({ address: sampleGoContract.address })`?

Reply:
The addresses used in[ the logs are provided by the EVM](https://github.com/loomnetwork/loomchain/blob/nestedlogs/evm/loomevm.go#L163),  and not under our control. EVM knows nothing about Go-Contracts so will only return the account number for EVM-contracts.
eth_getLogs and similar search for logs using the data in them (EVM-account#, Height, Topics) so knows nothing about tx-hashes or G0-transactions.
I did not use `getPastLogs` to get the logs. I used `getTransactionReceipt
` using the transaction hash. It did not seem important where I got the logs as I was not testing the getPastLogs functionality, just making sure the right logs where genearted.
I did use `getPastLogs` to get the transaction hash of the go-contract call so that I could get the receipt with the logs. There is no easy way to get the transaction hash of go-contracts. It is not returned as the transaction hash, and the web3 functionality does not understand go-contracts. (Although a go-transaction's tx-hash is the tendermint tx-hash so can in principle be calculated - maybe make a js utility function for this?)